src/test/java/uk/gov/ch/oauth/OAuthCoordinatorTest.java

```diff
@@ -0,0 +1,166 @@
+
+            Map<String, String> params = new HashMap<>();
+            params.put("state", "State");
+            params.put("code", "Code");
+
+            assertThrows(UnauthorisedException.class,
+                    () -> oAuthCoordinator.getPostCallbackRedirectURL(mockResponse, params));
+
+            verify(mockOAuth).validate("State", "Code", mockResponse);
+        }
+
+        @Test
+        @DisplayName("Get Post Back Returns identity provider results")
+        void getPostCallbackRedirectReturnsRedirectURLWhenValuesAreCorrect()
+                throws UnauthorisedException {
+            doReturn("Redirect URL").when(mockIdentityProvider).getRedirectUriPage();
+            doReturn(mockIdentityProvider).when(oAuthCoordinator).getIdentityProvider();
+            doReturn(mockOAuth).when(oAuthCoordinator).getOAuth();
+            when(mockOAuth.validate(anyString(), anyString(), any(HttpServletResponse.class)))
+                    .thenReturn(true);
+
+            Map<String, String> params = new HashMap<>();
+            params.put("state", "State");
+            params.put("code", "Code");
+
+            String ret = oAuthCoordinator.getPostCallbackRedirectURL(mockResponse, params);
+
+            assertEquals("Redirect URL", ret);
+            verify(mockOAuth).validate("State", "Code", mockResponse);
+            verify(mockIdentityProvider).getRedirectUriPage();
+        }
+    }
+
+    @Nested
+    class lazyGetterTests {
+
+        @Test
+        @DisplayName("Creates an IOAuth object if one doesn't exist and returns the same one if it does.")
+        void testGetOAuth() {
+            IOauth auth = oAuthCoordinator.getOAuth();
+            assertNotNull(auth);
+            assertEquals(auth, oAuthCoordinator.getOAuth());
+        }
+
+        @Test
+        @DisplayName("Creates an IIdentityProvider object if one doesn't exist and returns the same one if it does.")
+        void testGetIdentityProvider() {
+            IIdentityProvider identityProvider = oAuthCoordinator.getIdentityProvider();
+            assertNotNull(identityProvider);
+            assertEquals(identityProvider, oAuthCoordinator.getIdentityProvider());
+        }
+
+        @Test
+        @DisplayName("Creates an Environment Reader object if one doesn't exist and returns the same one if it does.")
+        void testGetEnvironmentReader() {
+            EnvironmentReader environmentReader = oAuthCoordinator.getEnvironmentReader();
+            assertNotNull(environmentReader);
+            assertEquals(environmentReader, oAuthCoordinator.getEnvironmentReader());
+        }
+
+        @Test
+        @DisplayName("Creates an Session Factory object if one doesn't exist and returns the same one if it does.")
+        void testGetSessionFactory() {
+            SessionFactory sessionFactory = oAuthCoordinator.getSessionFactory();
+            assertNotNull(sessionFactory);
+            assertEquals(sessionFactory, oAuthCoordinator.getSessionFactory());
```

Comment:
These assertEquals seem like useless tests to me, we know they are equal as `sessionFactory` for example is the result of `oAuthCoordinator.getSessionFactory()` above.

Reply:
They would be different if getSessionFactory() created a new SessionFactory each time.

Reply:
Okay I see